src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -0,0 +1,247 @@
+/**
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTRDeviceConnectivityMonitor.h"
+#import "MTRLogging_Internal.h"
+#import "MTRUnfairLock.h"
+
+#import <Network/Network.h>
+#import <dns_sd.h>
+#import <os/lock.h>
+
+@interface MTRDeviceConnectivityMonitor ()
+- (void)handleResolvedHostname:(const char *)hostName port:(uint16_t)port error:(DNSServiceErrorType)error;
+@end
+
+@implementation MTRDeviceConnectivityMonitor {
+    NSString * _instanceName;
+    DNSServiceRef _resolver;
+    NSMutableDictionary<NSString *, nw_connection_t> * _connections;
+
+    MTRDeviceConnectivityMonitorHandler _monitorHandler;
+    dispatch_queue_t _handlerQueue;
+}
+
+namespace {
+constexpr char kLocalDot[] = "local.";
+constexpr char kOperationalType[] = "_matter._tcp";
+}
+
+static dispatch_once_t sConnecitivityMonitorOnceToken;
+static os_unfair_lock sConnectivityMonitorLock;
+static NSUInteger sConnectivityMonitorCount;
+static DNSServiceRef sSharedResolverConnection;
+static dispatch_queue_t sSharedResolverQueue;
+
+- (instancetype)initWithInstanceName:(NSString *)instanceName
+{
+    if (self = [super init]) {
+        dispatch_once(&sConnecitivityMonitorOnceToken, ^{
+            sConnectivityMonitorLock = OS_UNFAIR_LOCK_INIT;
+        });
+        _instanceName = [instanceName copy];
+        _connections = [NSMutableDictionary dictionary];
+    }
+    return self;
+}
+
+- (void)dealloc
+{
+    if (_resolver) {
+        DNSServiceRefDeallocate(_resolver);
+    }
+}
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"<MTRDeviceConnectivityMonitor: %@>", _instanceName];
+}
+
++ (DNSServiceRef)_sharedResolverConnection
+{
+    os_unfair_lock_assert_owner(&sConnectivityMonitorLock);
+
+    if (!sSharedResolverConnection) {
+        DNSServiceErrorType dnsError = DNSServiceCreateConnection(&sSharedResolverConnection);
+        if (dnsError) {
+            MTR_LOG_ERROR("MTRDeviceConnectivityMonitor: DNSServiceCreateConnection failed %d", dnsError);
+            return NULL;
+        }
+        sSharedResolverQueue = dispatch_queue_create("MTRDeviceConnectivityMonitor", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
```

Comment:
Does this need to be a separate queue? Can we use the Matter work queue? Maybe there would be less dispatching / dispatch_sync required if we're already on the correct queue.

Reply:
Actually, given the requirement that the shared connection DNSServiceRef be dealloc'ed on the same queue set on it, I think keeping it separate avoids the issue of synchronizing cleanup when the matter stack is shut down.